flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java

```diff
@@ -0,0 +1,184 @@
+		List<RexNode> bottomProjects = bottomProgram.getProjectList()
+			.stream()
+			.map(bottomProgram::expandLocalRef)
+			.collect(Collectors.toList());
+
+		if (bottomProjects.size() != 1 || PythonUtil.isNonPythonCall(bottomProjects.get(0))) {
+			return false;
+		}
+
+		RexProgram topProgram = topCalc.getProgram();
+		List<RexNode> topProjects = topProgram.getProjectList()
+			.stream()
+			.map(topProgram::expandLocalRef)
+			.collect(Collectors.toList());
+		if (topProjects.size() != 1 || PythonUtil.isNonPythonCall(topProjects.get(0))) {
+			return false;
+		}
+
+		RexProgram middleProgram = middleCalc.getProgram();
+		if (middleProgram.getCondition() != null) {
+			return false;
+		}
+
+		List<RexNode> middleProjects = middleProgram.getProjectList()
+			.stream()
+			.map(middleProgram::expandLocalRef)
+			.collect(Collectors.toList());
+		int inputRowFieldCount = middleProgram.getInputRowType()
+			.getFieldList()
+			.get(0)
+			.getValue()
+			.getFieldList().size();
+
+		return isInputsCorrespondWithUpstreamOutputs(bottomProjects, middleProjects) &&
+			isFlattenCalc(middleProjects, inputRowFieldCount);
+	}
+
+	private boolean isInputsCorrespondWithUpstreamOutputs(List<RexNode> bottomProjects, List<RexNode> middleProjects) {
+		RexCall pythonCall = (RexCall) bottomProjects.get(0);
+		List<RexNode> pythonCallInputs = pythonCall.getOperands();
+		if (pythonCallInputs.size() != middleProjects.size()) {
+			return false;
+		}
+		for (int i = 0; i < pythonCallInputs.size(); i++) {
+			RexNode input = pythonCallInputs.get(i);
+			if (input instanceof RexInputRef) {
+				if (((RexInputRef) input).getIndex() != i) {
+					return false;
+				}
+			} else {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	private boolean isFlattenCalc(List<RexNode> middleProjects, int inputRowFieldCount) {
+		if (inputRowFieldCount != middleProjects.size()) {
+			return false;
+		}
+		for (int i = 0; i < inputRowFieldCount; i++) {
+			RexNode middleProject = middleProjects.get(i);
+			if (middleProject instanceof RexFieldAccess) {
+				RexFieldAccess rexField = ((RexFieldAccess) middleProject);
+				if (rexField.getField().getIndex() != i) {
+					return false;
+				}
+				RexNode expr = rexField.getReferenceExpr();
+				if (expr instanceof RexInputRef) {
+					if (((RexInputRef) expr).getIndex() != 0) {
+						return false;
+					}
+				} else {
+					return false;
+				}
+			} else {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	@Override
+	public void onMatch(RelOptRuleCall call) {
+		FlinkLogicalCalc bottomCalc = call.rel(0);
+		FlinkLogicalCalc middleCalc = call.rel(1);
+		FlinkLogicalCalc topCalc = call.rel(2);
+
+		RexProgram bottomProgram = bottomCalc.getProgram();
+		List<RexCall> bottomProjects = bottomProgram.getProjectList()
+			.stream()
+			.map(bottomProgram::expandLocalRef)
+			.map(x -> (RexCall) x)
+			.collect(Collectors.toList());
+		RexCall bottomPythonCall = bottomProjects.get(0);
```

Comment:
Should we also check whether there is condition in the bottomCalc and topCalc?

Reply:
`bottomCalc` and `topCalc` are PythonCalc which won't contain condition. so we don't need to check condition.